### DISCUSSION TITLE
Improving responsiveness of the analysis view

### DISCUSSION THREAD

**Discussion by pastewka** [ContactEngineering/topobank#760] [Ideas]
Analyses can take a long time to load, in particular of the digital surface twin have many measurements (as for our UNCD twin). I have a proposal to improve the responsiveness, that should also work for other parts of the system.

The reason for the long delays is that the Django backend aggregates *all* the data before dumping it to the client. This blocks the Django server an the user does not see anything before that process is finished.

Rather than storing the result dictionary of the analysis function in the database, we store it (as JSON) in the S3 system. When loading an analysis plot we create the plot but do not load the data. We then issues a bunch of (aynschronous) AJAX requests to load the respective JSON from the S3 and update the plot in a callback. The user should then see the data update in realtime. We should add a spinner to the top bar of the card to indicate that data is still loading.

This has the advantage that *all* data now comes from the S3 which should be able to handle many simultaneous requests more gracefully than our Django server. The Django backend is then just orchestrating the data collection but does not distribute the data.

I think this should require minimal changes to the code. If we want, we could even at a later stage move the creation of the Bokeh plot to JS (but that is not necessary and should be discussed separately).

Any thoughts, @mcrot ?

**Comment by mcrot** [comment]
I think it's a good idea and would be an great improvement for the usability.
Some things I would like to discuss before we start:

1. If we don't move all the plotting to JS in the beginning, the plotting still has to be done in Django; so every time one AJAX call returns we would call a Django view as callback. I'm not sure how well this works if multiple callbacks update the same part of the DOM at a similar time.
2. How we would display the progress bars then? On an extra tab? It would be good to still give feedback which analyses are still running.

**Reply by pastewka** [reply]
Regarding these points:

1. This would be a single call generating the Django view (i.e. plot HTML) without any data. Data should then be loaded to from a separate endpoint (i.e. S3 directly). Not sure that this needs direct update of the DOM, it need update of the Bokeh plot. Maybe even `AjaxDataSource` could be used to schieve this.

2. That's right. Maybe we could stick this into the "Task information" modal that all analyses have. I would indicate with a spinner that things are going on in the background but without giving detail on what is still running.
